pkg/kv/kvserver/store_snapshot.go

```diff
@@ -362,23 +397,37 @@ func (kvSS *kvBatchSnapshotStrategy) Send(
 	// CRDB, as of VersionUnreplicatedTruncatedState).
 	bytesSent := int64(0)
 
-	// Iterate over all keys using the provided iterator and stream out batches
-	// of key-values.
-	kvs := 0
+	// Iterate over all keys (point keys and range keys) using the provided
+	// iterators and stream out batches of key-values.
+	var kvs, rangeKVs int
 	var b storage.Batch
 	defer func() {
 		if b != nil {
 			b.Close()
 		}
 	}()
-	for iter := snap.Iter; ; iter.Next() {
-		if ok, err := iter.Valid(); err != nil {
-			return 0, err
-		} else if !ok {
-			break
+
+	flushBatch := func() error {
+		if err := kvSS.sendBatch(ctx, stream, b); err != nil {
+			return err
 		}
+		bLen := int64(b.Len())
+		bytesSent += bLen
+		recordBytesSent(bLen)
+		b.Close()
+		b = nil
+		return nil
+	}
+
+	iter := snap.PointKeyIter
```

Comment:
Do we need to mention somewhere the order in which we put data in batches? Receive method currenly lists its expectations about the order of keys. With range keys added, we will be emitting all point keys for all subranges that iterator gives us, then all range keys for all subranges. In practice we now only have range keys in the replicated user keys, but if something changes in future it would be good to know upfront that this could potentially break.

Reply:
Thanks for bringing this up. There's a problem here in that `multiSSTWriter` requires each key range to be processed separately.

Reply:
Come to think of it, there's an additional problem in that not doing so would created extremely wide SSTs, that could cause problems for Pebble.

Reply:
This ended up being pretty annoying. I went back and forth on a few different approaches, but the solution I landed on for now was to change `ReplicaEngineDataIterator` to iterate over both point keys and range keys, but emit them in the desired order. That is, instead of having separate point/range key iterators and trying to coordinate them to emit keys with some reasonable grouping, it will instead, for each key range, first emit all point keys and then emit all range keys. This allows both efficient emission of keys and efficient ingestion of them.

I'm not convinced that this is the optimal approach, but it gets the job done. I added a `TODO` to revisit this later.

Would appreciate if you could have another look tomorrow morning.